### PR TITLE
(feat) support datasource delete by name api

### LINKED ISSUE #7476: [Feature request] Ability to delete datasources by name
Please include this information:
- **What Grafana version are you using?** Grafana v4.1.0 (commit: 98b648d) 
- **What datasource are you using?** InfluxDB, Elasticsearch
- **What OS are you running grafana on?** Ubuntu 14.04
- **What did you do?** Trying to delete a datasource by name, e.g., `DELETE /api/datasources/:datasourceName`
- **What was the expected result?** The datasource is deleted
- **What happened instead?** An error is returned

This kind of functionality would be useful for automation, when datasources need to be created and removed depending on other dependencies.

### PR DESCRIPTION
**Pull request by huydx**
Fix https://github.com/grafana/grafana/issues/7476

### CODE REVIEW

File: pkg/api/api.go [hunk 1]
```diff
@@ -198,7 +198,7 @@ func (hs *HttpServer) registerRoutes() {
 			r.Get("/", GetDataSources)
 			r.Post("/", quota("data_source"), bind(m.AddDataSourceCommand{}), AddDataSource)
 			r.Put("/:id", bind(m.UpdateDataSourceCommand{}), wrap(UpdateDataSource))
-			r.Delete("/:id", DeleteDataSource)
+			r.Delete("/:idOrName", DeleteDataSource)
```
Comment (bergquist) [inline pkg/api/api.go]: I think this url should follow the same pattern as `r.Get("/name/:name")`. Using the same handler for both name and id makes it impossible to use a number as name.

File: pkg/api/datasources.go [hunk 2]
```diff
@@ -69,16 +71,39 @@ func GetDataSourceById(c *middleware.Context) Response {
 }
 
 func DeleteDataSource(c *middleware.Context) {
```
Comment (bergquist) [inline pkg/api/datasources.go]: Since we should have two urls for deleting datasources we should also splitt this handler into two.

**Review by bergquist — CHANGES_REQUESTED** [review]

**Comment by bergquist** [comment]
Great work! just need a few changes before we merge this.

**Comment by huydx** [comment]
fix

**Review by bergquist — APPROVED** [review]

**Comment by bergquist** [comment]
Great work! :)

**Pull request merged**